main/src/com/google/refine/expr/functions/ToDate.java

```diff
@@ -81,52 +82,65 @@ public Object call(Properties bindings, Object[] args) {
-                    return new EvalError("Invalid non-string format argument " + args[i].toString());
+                int startIndex = 1; // set this as where to start picking formats from - the first argument
+                if(args[1] instanceof Boolean) {
+                    monthFirst = (Boolean) args[1];
+                    startIndex = 2; // if the second argument is a boolean, then start picking formats from the third argument
                 }
-                formats.add(StringUtils.trim((String) args[i]));
-            }
-            try {
-                if (month_first != null) {
-                    date = parse(o1, month_first, formats);
-                } else {
-                    date = parse(o1, formats);
+                for (int i = startIndex; i < args.length; i++) {
+                    if (!(args[i] instanceof String)) {
+                        // if the argument is not a string, return an error
+                        return new EvalError("Invalid non-string format argument " + args[i].toString());
+                    }
+                    // add the format to the list of formats
+                    formats.add(StringUtils.trim((String) args[i]));
                 }
-            } catch (DateFormatException e) {
-                return new EvalError(e.getMessage());
             }
         }
         if(date != null) {
+            // return the parsed date if it was success
             return date;
+        } else {
+            try {
+                monthFirst = monthFirst == null || monthFirst;
+
+                // parse the date with the user defined month_first and the formats
+                date = parse(o1, monthFirst, formats); // checking ig the month_first is null is already done in the parse method
+            } catch (DateFormatException | DateTimeParseException e) {
+                return new EvalError(e.getMessage());
+            }
         }
-        return new EvalError("Unable to convert to a date");
+        return date == null ? new EvalError("Unable to convert to a date") : date;
     }
     
-    private OffsetDateTime parse(String o1, Boolean month_first, List<String> formats) throws DateFormatException {
-        if(month_first != null) {
+    private OffsetDateTime parse(String o1, Boolean monthFirst, List<String> formats) throws DateFormatException {
+        if(monthFirst != null && formats.size() == 0) {
             try {
-               return CalendarParser.parseAsOffsetDateTime( o1, (month_first) ? CalendarParser.MM_DD_YY : CalendarParser.DD_MM_YY);
-            } catch (CalendarParserException e) {
+               return CalendarParser.parseAsOffsetDateTime( o1, (monthFirst) ? CalendarParser.MM_DD_YY : CalendarParser.DD_MM_YY);
+            } catch (CalendarParserException ignored) {
            }
         }
-        return parse(o1,formats);
+        OffsetDateTime date = parse(o1, formats);
+        if (Boolean.FALSE.equals(monthFirst)) {
```

Comment:
@elroykanye What test covers this fully?

Reply:
https://github.com/OpenRefine/OpenRefine/blob/master/main/tests/server/src/com/google/refine/expr/functions/ToDateTests.java#L76-L78

It's actually what has kept me on hold for some time 😃  I noticed that formats were not in the tests when `monthFirst` was explicitly stated.

Reply:
OK, well, I think we need to be descriptive in our tests (to cover our assumptions of formats) so that in the future, in case we make Date library changes, or the JDK itself goes crazy, then at least our test cases are described well with comments, the formats described so eliminate ambiguity for other developers in the future, and basically make OpenRefine Date parsing a bit more bulletproof with robust testing here.

Is your intention to improve the tests in a few more commits on this?

Reply:
I did plan to improve the tests for cases where just 1 or 2 args are provided which is the case handled by the introduced fix. I thought it wise to leave the other tests as they are just so I am sure the improvements do not break any uses of toDate()